### ISSUE TITLE
Server-side pagination does not work (exception)

### ISSUE THREAD

**Issue by DominicTobias** [ag-grid/ag-grid#3419]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of ag-Grid'
[] support request => see 'Requesting Community Support'
```

**Customers of ag-Grid**

Yes (UBS Neo)

**Providing a Reproducible Scenario**

 "ag-grid-community": "^21.2.1",
 "ag-grid-enterprise": "^21.2.1",
 "ag-grid-react": "^21.2.1",
 "react": "^16.8.6",

Creating a very simple serverside pagination:

```js
const serverSideDatasource = {
  async getRows(params) {
    try {
      const { request, successCallback } = params;
      const { total, issues } = await jiraService.issues(request.startRow, 50);
      successCallback(issues, total);
    } catch (err) {
      console.error(err);
      params.failCallback();
    }
  },
};

<DataGrid
  data={[]}
  columnDefs={columnDefs}
  rowModelType="serverSide"
  serverSideDatasource={serverSideDatasource}
  pagination
  paginationPageSize={50}
/>
```

Loads first 50 items fine but throws the exception when trying to go to the next page, in this case `getRows` is NOT called, exception happens internally before this point.

Data returns 50 items and the total number of items in the database.

![pagination-error](https://user-images.githubusercontent.com/760314/66036857-3566ba00-e506-11e9-8dd1-98c79275865f.png)

```
TypeError: Cannot read property 'rowIndex' of undefined
    at ServerSideCache.getTopLevelRowDisplayedIndex (serverSideCache.js:364)
    at ServerSideRowModel.getTopLevelRowDisplayedIndex (serverSideRowModel.js:486)
    at PaginationProxy.calculatePagesMasterRowsOnly (paginationProxy.js:334)
    at PaginationProxy.calculatePages (paginationProxy.js:271)
    at PaginationProxy.onModelUpdated (paginationProxy.js:106)
    at eventService.js:189
    at Array.forEach (<anonymous>)
    at Function.Utils.forEachSnapshotFirst (general.js:1642)
    at EventService.dispatchToListeners (eventService.js:183)
    at EventService.dispatchEvent (eventService.js:153)
    at ServerSideRowModel.onCacheUpdated (serverSideRowModel.js:424)
    at eventService.js:189
    at Array.forEach (<anonymous>)
    at Function.Utils.forEachSnapshotFirst (general.js:1642)
    at EventService.dispatchToListeners (eventService.js:183)
    at EventService.dispatchEvent (eventService.js:153)
    at ServerSideCache.BeanStub.dispatchEvent (beanStub.js:105)
    at ServerSideCache.RowNodeCache.onCacheUpdated (rowNodeCache.js:292)
    at ServerSideCache.RowNodeCache.checkVirtualRowCount (rowNodeCache.js:185)
    at ServerSideCache.RowNodeCache.onPageLoaded (rowNodeCache.js:109)
    at eventService.js:189
    at Array.forEach (<anonymous>)
    at Function.Utils.forEachSnapshotFirst (general.js:1642)
    at EventService.dispatchToListeners (eventService.js:183)
    at EventService.dispatchEvent (eventService.js:153)
    at ServerSideBlock.BeanStub.dispatchEvent (beanStub.js:105)
    at ServerSideBlock.RowNodeBlock.pageLoaded (rowNodeBlock.js:263)
    at _callee$ (index.js:85)
    at tryCatch (runtime.js:45)
    at Generator.invoke [as _invoke] (runtime.js:264)
    at Generator.prototype.<computed> [as next] (runtime.js:98)
    at asyncGeneratorStep (asyncToGenerator.js:3)
    at _next (asyncToGenerator.js:25)
console.<computed> @ index.js:1437
_callee$ @ index.js:87
tryCatch @ runtime.js:45
invoke @ runtime.js:264
prototype.<computed> @ runtime.js:98
asyncGeneratorStep @ asyncToGenerator.js:3
_next @ asyncToGenerator.js:25
Promise.then (async)
asyncGeneratorStep @ asyncToGenerator.js:13
_next @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
getRows @ index.js:69
(anonymous) @ serverSideBlock.js:277
setTimeout (async)
ServerSideBlock.loadFromDatasource @ serverSideBlock.js:275
RowNodeBlock.load @ rowNodeBlock.js:197
RowNodeBlockLoader.performCheckBlocksToLoad @ rowNodeBlockLoader.js:103
RowNodeBlockLoader.checkBlockToLoad @ rowNodeBlockLoader.js:79
RowNodeCache.checkBlockToLoad @ rowNodeCache.js:176
RowNodeCache.postCreateBlock @ rowNodeCache.js:161
ServerSideCache.createBlock @ serverSideCache.js:376
ServerSideCache.getRow @ serverSideCache.js:318
ServerSideRowModel.getRow @ serverSideRowModel.js:455
PaginationProxy.getRow @ paginationProxy.js:146
RowRenderer.createOrUpdateRowComp @ rowRenderer.js:928
(anonymous) @ rowRenderer.js:835
RowRenderer.redraw @ rowRenderer.js:834
RowRenderer.redrawAfterModelUpdate @ rowRenderer.js:467
RowRenderer.registerGridComp @ rowRenderer.js:149
GridPanel.init @ gridPanel.js:229
(anonymous) @ context.js:269
(anonymous) @ context.js:268
Context.forEachMetaDataInHierarchy @ context.js:200
(anonymous) @ context.js:261
Context.callLifeCycleMethods @ context.js:260
Context.wireBeans @ context.js:97
Context.wireBean @ context.js:84
Context.createComponentFromElement @ context.js:72
(anonymous) @ component.js:100
Component.createChildComponentsFromTags @ component.js:99
(anonymous) @ component.js:118
Component.createChildComponentsFromTags @ component.js:99
Component.setTemplateFromElement @ component.js:172
Component.setTemplate @ component.js:160
GridCore.init @ gridCore.js:103
(anonymous) @ context.js:269
(anonymous) @ context.js:268
Context.forEachMetaDataInHierarchy @ context.js:200
(anonymous) @ context.js:261
Context.callLifeCycleMethods @ context.js:260
Context.wireBeans @ context.js:97
Context.wireBean @ context.js:84
Grid @ grid.js:334
AgGridReact.componentDidMount @ agGridReact.js:124
commitLifeCycles @ react-dom.development.js:21169
commitLayoutEffects @ react-dom.development.js:24165
callCallback @ react-dom.development.js:363
invokeGuardedCallbackDev @ react-dom.development.js:412
invokeGuardedCallback @ react-dom.development.js:466
commitRootImpl @ react-dom.development.js:23930
unstable_runWithPriority @ scheduler.development.js:676
runWithPriority$2 @ react-dom.development.js:11855
commitRoot @ react-dom.development.js:23750
runRootCallback @ react-dom.development.js:22836
(anonymous) @ react-dom.development.js:11907
unstable_runWithPriority @ scheduler.development.js:676
runWithPriority$2 @ react-dom.development.js:11855
flushSyncCallbackQueueImpl @ react-dom.development.js:11902
flushSyncCallbackQueue @ react-dom.development.js:11890
scheduleUpdateOnFiber @ react-dom.development.js:22694
enqueueSetState @ react-dom.development.js:13731
push../node_modules/react/cjs/react.development.js.Component.setState @ react.development.js:324
(anonymous) @ DynamicImport.js:12
Promise.then (async)
componentDidMount @ DynamicImport.js:11
commitLifeCycles @ react-dom.development.js:21169
commitLayoutEffects @ react-dom.development.js:24165
callCallback @ react-dom.development.js:363
invokeGuardedCallbackDev @ react-dom.development.js:412
invokeGuardedCallback @ react-dom.development.js:466
commitRootImpl @ react-dom.development.js:23930
unstable_runWithPriority @ scheduler.development.js:676
runWithPriority$2 @ react-dom.development.js:11855
commitRoot @ react-dom.development.js:23750
scheduleUpdateOnFiber @ react-dom.development.js:22683
scheduleRootUpdate @ react-dom.development.js:25713
updateContainerAtExpirationTime @ react-dom.development.js:25739
updateContainer @ react-dom.development.js:25839
(anonymous) @ react-dom.development.js:26397
unbatchedUpdates @ react-dom.development.js:22979
legacyRenderSubtreeIntoContainer @ react-dom.development.js:26396
render @ react-dom.development.js:26487
./src/index.js @ index.js:11
__webpack_require__ @ bootstrap:790
fn @ bootstrap:150
0 @ history.js:3
__webpack_require__ @ bootstrap:790
checkDeferredModules @ bootstrap:45
webpackJsonpCallback @ bootstrap:32
(anonymous) @ main.chunk.js:1
Show 42 more frames
```

**Comment by DominicTobias** [comment]
This issue happens do to this line:

```js
var lastRowNode = blockBefore_1.getRowUsingLocalIndex(lastRowTopLevelIndex, true);
```

For some reason you calc that lastRowTopLevelIndex is `199` here:

```js
var lastRowTopLevelIndex = blockSize * (blockId + 1) - 1;
```

(blockSize is 100)

The last number that returns something is in fact `99`, this works:

```js
blockBefore_1.getRowUsingLocalIndex(99, true);
```

**Comment by DominicTobias** [comment]
It looks like you fixed this bug in 21.2.2
